File: internal/state/state.go
package state

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type State struct {
	Bucket string `json:"bucket,omitempty"`
	Prefix string `json:"prefix,omitempty"`

	path string
}

func DefaultPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".config", "yelo", "state.json")
}

func Load(path string) (*State, error) {
	if path == "" {
		path = DefaultPath()
	}

	s := &State{path: path}

	data, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return s, nil
	}
	if err != nil {
		return nil, fmt.Errorf("reading state: %w", err)
	}

	if err := json.Unmarshal(data, s); err != nil {
		return nil, fmt.Errorf("parsing state: %w", err)
	}
	s.path = path
	return s, nil
}

func (s *State) Save() error {
	dir := filepath.Dir(s.path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("creating state dir: %w", err)
	}

	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling state: %w", err)
	}

	return os.WriteFile(s.path, data, 0o644)
}

func (s *State) SetBucket(bucket string) {
	s.Bucket = bucket
	s.Prefix = ""
}

File: src/state.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct State {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub bucket: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub prefix: String,
}

impl State {
    pub fn path() -> PathBuf {
        dirs::home_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join(".yelo/state.json")
    }

    pub fn load() -> Result<Self> {
        let path = Self::path();
        if !path.exists() {
            return Ok(Self::default());
        }
        let data = fs::read_to_string(&path).context("reading state")?;
        let st: State = serde_json::from_str(&data).context("parsing state")?;
        Ok(st)
    }

    pub fn save(&self) -> Result<()> {
        let path = Self::path();
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent).context("creating state dir")?;
        }
        let data = serde_json::to_string_pretty(self).context("serializing state")?;
        fs::write(&path, data).context("writing state")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_state_empty() {
        let st = State::default();
        assert!(st.bucket.is_empty());
        assert!(st.prefix.is_empty());
    }

    #[test]
    fn serde_round_trip() {
        let st = State {
            bucket: "my-bucket".into(),
            prefix: "photos/2024/".into(),
        };
        let json = serde_json::to_string_pretty(&st).unwrap();
        let parsed: State = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.bucket, "my-bucket");
        assert_eq!(parsed.prefix, "photos/2024/");
    }

    #[test]
    fn serde_empty_fields_omitted() {
        let st = State::default();
        let json = serde_json::to_string(&st).unwrap();
        assert!(!json.contains("bucket"));
        assert!(!json.contains("prefix"));
    }

    #[test]
    fn serde_partial_fields() {
        let json = r#"{"bucket": "test"}"#;
        let st: State = serde_json::from_str(json).unwrap();
        assert_eq!(st.bucket, "test");
        assert!(st.prefix.is_empty());
    }
}
